chains/notifier.go

```diff
@@ -0,0 +1,33 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package chains
+
+import (
+	"context"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+)
+
+type ChangeNotifier struct {
+	block.ChainVM
+	OnChange func()
+}
```

Comment:
While this method of implementing a VM to intercept the calls is nice... I don't think this correctly handles the optional interfaces that VMs may implement.

`ChangeNotifier` will only implement `block.ChainVM` even if the `block.ChainVM` that it wraps implements other optional VM interfaces (such as state sync and batched operations).

As an example, I don't think that the C-chain will be able to state-sync with this wrapping.